### DISCUSSION TITLE
Data Fetching At Very First Undefined Issue

### DISCUSSION THREAD

**Discussion by EhanDuan** [vercel/next.js#38139] [Help] [answered]
## Overview
Hi, I am using `nextjs`, `firebase` to build an webapp.If you want to directly go to the problem, please scroll down to **Idea No.3** and **Question part.** 

## Senario
In a page, it will show the current user's `focus skills` among all `skills`.  The result will be **represented as checkboxes**.  It will also allow user to change `focus skills`;

**User flow:**
+ At first, when user request pages, the user will see his previously selected skills already checked in the checkboxes
+ Then, the user checks other checkboxes. Click confirm.
+ Still in the same page but the result is updated.

## Idea
1. I need **all skills at requesting time**, so I use `getServerSideProps` to fetch all skills
```ts
export async function getServerSideProps() {
  const querySnapshot = await getDocs(collection(db, "skills"));
  const skillsData: any[] = [];
  const sIds: string[] = [];

  querySnapshot.forEach((doc) => {
    skillsData.push(doc.data());
    sIds.push(doc.id);
  });

  return { props: { skillsData, sIds } };
}
```

2. It is a **user-specific page,** I need to fetch user Data. Here I fetch in the page component itself. (I try to fetch user data  in `getServersideProps`, but it failed due to user initial is undefined complained by firebase)
```ts
const FocusList = ({ skillsData, sIds }) => {
  // get user infos
  const user = useAuth().user;

  // get user current focus skills and set Id
  const [userData, loading, error] = useDocument(doc(db, "users", user.uid), {
    snapshotListenOptions: { includeMetadataChanges: true }
  });
  // other case...
```

3. **(Problem Appears in This Step)** Then, I use `useState` with `selectionMap` intial value. (Map<string, boolean>), key is **id** of `skill`, and value is true if it is checked. The inital value of map is based on all skills and user focus skills.
```ts
 // 1. added previous selected
  const selectionMap = new Map<string, boolean>();
 // Reason for prev... is for checkbox group, default value should be array. 
  const prevSelectedIds: string[] = [];

  for (const sId of userData?.data()?.focus) {
    selectionMap.set(sId, true);
    prevSelectedIds.push(sId);
  }
  // 2. add other skills with false
  // traverse
  for (let i = 0; i < sIds.length; i++) {
    if (selectionMap.has(sIds[i])) continue;
    selectionMap.set(sIds[i], false);
  }

  // create a hashmap as state, with all sIds as key, and true/false as value
  const [selectedMap, setSelectedMap] = useState(selectionMap);
```

## Problem
If I start the project, It will complain the following error.
<img width="833" alt="image" src="https://user-images.githubusercontent.com/75397936/176345579-b8e84213-8949-4fbd-ad66-9a39e79d017e.png">
If I comment out part of Idea No.3 code like following. There is no error. But the data is empty.
```ts
  // for (const sId of userData?.data()?.focus) {
  //   selectionMap.set(sId, true);
  //   prevSelectedIds.push(sId);
  // }
```
<img width="1034" alt="image" src="https://user-images.githubusercontent.com/75397936/176346800-461ee526-278d-4b05-a5c3-e861b50e880a.png">

In this way, I think the error is due to **for loop of data from firebase** .
```ts
for (const sId of userData?.data()?.focus)
```

I can make sure the `useData`, `userData.data()`and `useData.data().focus` exist if network okay. **But it seems like at the very first beginning**,  `userData?.data()?.focus` **is undefined from error image**. So the code here will recognize as undefined even it will eventually get the data.

My question is: how to handle this error? I was thinking use `hook` which is naturally to update the data. But it still cannot prevent the starting point undefined. 

I also have similar questions which are all related to the initial undefined errors.

**In addition, I want to ask questions efficiently, do you have any suggestions about the way I am asking? Like too unnecessary information or "Great to offer the context of your problem". Anything helps!**

## Code (Too long)
```tsx
const FocusList = ({ skillsData, sIds }) => {
  // get user infos
  const user = useAuth().user;

  // get user current focus skills and set Id
  const [userData, loading, error] = useDocument(doc(db, "users", user.uid), {
    snapshotListenOptions: { includeMetadataChanges: true }
  });

  // 1. added previous selected skills Ids
  const selectionMap = new Map<string, boolean>();
  const prevSelectedIds: string[] = [];

  // for (const sId of userData?.data()?.focus) {
  //   selectionMap.set(sId, true);
  //   prevSelectedIds.push(sId);
  // }

  // 2. add other skills Id with false
  for (let i = 0; i < sIds.length; i++) {
    if (selectionMap.has(sIds[i])) continue;
    selectionMap.set(sIds[i], false);
  }

  // create a hashmap as state, with all sIds as key, and true/false as value
  const [selectedMap, setSelectedMap] = useState(selectionMap);

  // create a roaster for Id to Name
  const skillsRoaster = useMemo(() => {
    const map = new Map<string, string>();

    sIds.forEach((sId: string, index: number) => {
      map.set(sId, skillsData[index].name);
    });
    return map;
    // eslint-disable-next-line
  }, [sIds, skillsData]);

  const handleCheck = (e) => {
    e.preventDefault();
    setSelectedMap(selectedMap);
  };

  const handleSubmit = async () => {
    const selectedList: String[] = [];
    for (const [skill, checked] of selectedMap) {
      if (checked) {
        selectedList.push(skill);
      }
    }
    if (selectedList.length < 1) {
      alert("You need to select at least 1 skill");
    } else {
      if (user) {
        await setDoc(
          doc(db, "users", user.uid),
          { focus: selectedList },
          { merge: true }
        );
        alert("User focus skills updated successfully");
      }
    }
  };

  return (
    <Flex flexDirection="column" flexFlow="row wrap">
      <Navbar />
      <Sidebar />
      <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>
        <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>

          <CheckboxGroup colorScheme="blue" defaultValue={prevSelectedIds}>
            <Wrap>
              {sIds.length > 0 &&
                sIds.map((sId, index) => (
                  <Button
                    as={Checkbox}
                    key={sId}
                    value={sId}
                    onChange={handleCheck}
                    isChecked={selectedMap.get(sId)}
                  >
                    {skillsRoaster.get(sId)}
                  </Button>
                ))}
            </Wrap>
          </CheckboxGroup>
          <Button>
            Confirm
          </Button>
        </Box>
      </Box>
    </Flex>
  );
};

export async function getServerSideProps() {
  const querySnapshot = await getDocs(collection(db, "skills"));
  const skillsData: any[] = [];
  const sIds: string[] = [];

  querySnapshot.forEach((doc) => {
    skillsData.push(doc.data());
    sIds.push(doc.id);
  });

  return { props: { skillsData, sIds } };
}

export default withProtected(FocusList);

```

**Comment by icyJoseph** [comment, marked as answer]
Assuming there's no way to get user data inside `getServerSideProps`(GSSP). And also leaving my coding style preferences aside.

We have 2 frames to think about, first frame renders with inputs from `GSSP`, and runs `useDocument`, I'll assume that `useAuth` somehow has the user from the first frame, but it might not be the case.

In this first frame, an async job is started by `useDocument` to go and get the data for `userData`, during this time, I assuming, loading is `true`, and error, `null` | `undefined`.

One way to fix this is to assume the cost of the async work through a `useEffect`:

```ts
  useEffect(() => {
    if (loading) return;
    if (error) return;

    // here we assume `userData`and `sIds`to be stable, meaning, they won't change through renders
    const selectionMap = new Map<string, boolean>();
    const prevSelectedIds: string[] = [];

    for (const sId of userData?.data()?.focus) {
      selectionMap.set(sId, true);
      prevSelectedIds.push(sId);
    }

    for (const sId of sIds) {
      if (selectionMap.has(sId)) continue;
      selectionMap.set(sId, false);
    }

    setSelectedMap(selectionMap);
    setInitiallySelected(prevSelectedIds);
  }, [error, loading, userData, sIds]);
```

So that the final code looks like:

```tsx
const FocusList = ({ skillsData, sIds }) => {
  // get user infos
  const user = useAuth().user;

  // get user current focus skills and set Id
  const [userData, loading, error] = useDocument(doc(db, "users", user.uid), {
    snapshotListenOptions: { includeMetadataChanges: true },
  });

  // create a hashmap as state, with all sIds as key, and true/false as value
  const [selectedMap, setSelectedMap] = useState(new Map<string, boolean>());

  const [initiallySelected, setInitiallySelected] = useState<string[]>([]);

  useEffect(() => {
    if (loading) return;
    if (error) return;

    // here we assume `userData`and `sIds`to be stable, meaning, they won't change through renders
    const selectionMap = new Map<string, boolean>();
    const prevSelectedIds: string[] = [];

    for (const sId of userData?.data()?.focus) {
      selectionMap.set(sId, true);
      prevSelectedIds.push(sId);
    }

    for (const sId of sIds) {
      if (selectionMap.has(sId)) continue;
      selectionMap.set(sId, false);
    }

    setSelectedMap(selectionMap);
    setInitiallySelected(prevSelectedIds);
  }, [error, loading, userData, sIds]);

  // create a roaster for Id to Name
  const skillsRoaster = useMemo(() => {
    const map = new Map<string, string>();

    sIds.forEach((sId: string, index: number) => {
      map.set(sId, skillsData[index].name);
    });
    return map;
    // eslint-disable-next-line
  }, [sIds, skillsData]);

  const handleCheck = (e) => {
    e.preventDefault();
    setSelectedMap(selectedMap);
  };

  const handleSubmit = async () => {
    const selectedList: String[] = [];
    for (const [skill, checked] of selectedMap) {
      if (checked) {
        selectedList.push(skill);
      }
    }
    if (selectedList.length < 1) {
      alert("You need to select at least 1 skill");
    } else {
      if (user) {
        await setDoc(
          doc(db, "users", user.uid),
          { focus: selectedList },
          { merge: true }
        );
        alert("User focus skills updated successfully");
      }
    }
  };

  return (
    <Flex flexDirection="column" flexFlow="row wrap">
      <Navbar />
      <Sidebar />
      <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>
        <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>
          <CheckboxGroup colorScheme="blue" defaultValue={initiallySelected}>
            <Wrap>
              {sIds.length > 0 &&
                sIds.map((sId, index) => (
                  <Button
                    as={Checkbox}
                    key={sId}
                    value={sId}
                    onChange={handleCheck}
                    isChecked={selectedMap.get(sId)}
                  >
                    {skillsRoaster.get(sId)}
                  </Button>
                ))}
            </Wrap>
          </CheckboxGroup>
          <Button>Confirm</Button>
        </Box>
      </Box>
    </Flex>
  );
};
```

I still think though, that it might not behave.

I am not sure if a Map is the most appropriate to handle these updates though. You'd need to return a new Map every time the user clicks on something to trigger rendering, since you are using controlled inputs.

Let me know how it goes! I am glad to help out.

**Reply by EhanDuan** [reply]
@icyJoseph Hi, Thanks again for helping!  Your solution is clear. Using an empty map or array and combined with `useEffect` could avoid the first frame's undefined issue. I used your code. **I think there is still some bug in my code.**

### `initiallySelected` in Return is Empty at First
When the page is requested, all the checkboxes are not checked.  I console log the `initiallySelected` and `selectedMap` exactly before the return.  Here are the image of box and two variables.
<img width="1790" alt="image" src="https://user-images.githubusercontent.com/75397936/176470840-5ad861dc-797e-4adb-91a0-8132513e95b6.png">

From the running time, I think the reason is at the first frame, 2 variables are empty ( checkboxes use this empty data) . Then, their values update in the `useEffect` block. But **this does not trigger rerendering as expected for some reason**, even 2 variables are using `useState`. 

#### Checking By Hard-coded String[]
To prove my thought, I change the `initiallySelected` into hard-coded array like this. It worked only checking the inital data.
```tsx
const [initiallySelected, setInitiallySelected] = useState<string[]>([
    "1YnW6LszVyBX8suIYeyv",
    "N2CxgcptTee745wGpIvy"
  ]);
```
#### TimeLine
To narrow down, I also console log the `prevSelected` here.
```tsx
  useEffect(() => {
    if (loading) return;
    // ...
    console.log(`At ${performance.now()} , the prevSelectedIds is :`);
    console.log(prevSelectedIds);
  }, [error, loading, userData, sIds]);
```
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/75397936/176476015-f05f0661-12b2-46e3-9c4a-3c8e81902612.png">
<img width="1213" alt="image" src="https://user-images.githubusercontent.com/75397936/176478632-744e990c-4f4b-4daf-b189-c196d466fd56.png">

It should have triggered `useState` rerendering at 5308. But why not?

## Others
### Get User Data in GSSP
From your assumptions, I feel I should or could get user's data in GSSP. Previously, I tried something like this with a similar error.
(Note, here auth is from `firebase` using `getAuth()`.
<img width="851" alt="image" src="https://user-images.githubusercontent.com/75397936/176480806-d458fbd4-6acf-4010-92d8-c40ae73fc2d5.png">

 The user is initially undefined. I checked online, the conclusion is: firebase auth needs roundtrip, which is the same as what we discussed above. I cannot use hook in GSSP.  

I have been thinking of using cookies session, but might implement in next version.

### Map Data structure
The purpose here is to store which `sId` is/are selected. What came in my mind, `hashSet<string>`, `hashmap<string, boolean>`, `array<string>`. All 3 could add in O(1). But for remove, array needs O(n) while previous 2 need O(1).  `includes` method is the same. Am I missing other aspects?

### Coding Style
**I would like to hear your suggestions!** This will helps me a lot in my career life!

**Reply by icyJoseph** [reply]
Mm right, the initial values are empty, in the first frame, so it is problem.

Let's try with this, a render props pattern. We'll wait with rendering the checkboxes, while the data is fetched, but we can render the rest in the meantime.

```tsx
import { ReactNode, useState, useEffect, useMemo, ReactElement } from "react";

type RenderWithInitiallySelectedProps = {
  focus: string[] | null | undefined;
  fallback?: NonNullable<ReactNode> | null;
  children: ({
    initiallySelected,
  }: {
    initiallySelected: string[];
  }) => ReactElement;
};

const RenderWithInitiallySelected = ({
  focus,
  fallback = null,
  children,
}: RenderWithInitiallySelectedProps) => {
  const [initiallySelected, setInitiallySelected] = useState<string[]>([]);

  const [status, setStatus] = useState<"pending" | "ready">("pending");

  useEffect(() => {
    if (status === "ready") return;
    if (!Array.isArray(focus)) return;

    setInitiallySelected(() => focus.slice(0));

    setStatus("ready");
  }, [status, focus]);

  if (status === "pending") return <>{fallback}</>;

  return children({ initiallySelected });
};
```

I think you could pass various fallback, including null, to show something while firebase stuff resolves from the network.

Then your component would look like:

```tsx
const FocusList = ({ skillsData, sIds }) => {
  // get user infos
  const user = useAuth().user;

  // get user current focus skills and set Id
  const [userData, loading, error] = useDocument(doc(db, "users", user.uid), {
    snapshotListenOptions: { includeMetadataChanges: true },
  });

  // create a hashmap as state, with all sIds as key, and true/false as value
  const [selectedMap, setSelectedMap] = useState(new Map<string, boolean>());

  useEffect(() => {
    if (loading) return;
    if (error) return;

    // here we assume `userData`and `sIds`to be stable, meaning, they won't change through renders
    const selectionMap = new Map<string, boolean>();

    for (const sId of userData?.data()?.focus) {
      selectionMap.set(sId, true);
    }

    for (const sId of sIds) {
      if (selectionMap.has(sId)) continue;
      selectionMap.set(sId, false);
    }

    setSelectedMap(selectionMap);
  }, [error, loading, userData, sIds]);

  // create a roaster for Id to Name
  const skillsRoaster = useMemo(() => {
    const map = new Map<string, string>();

    sIds.forEach((sId: string, index: number) => {
      map.set(sId, skillsData[index].name);
    });
    return map;
    // eslint-disable-next-line
  }, [sIds, skillsData]);

  const handleCheck = (e) => {
    e.preventDefault();
    setSelectedMap(selectedMap);
  };

  const handleSubmit = async () => {
    const selectedList: String[] = [];
    for (const [skill, checked] of selectedMap) {
      if (checked) {
        selectedList.push(skill);
      }
    }
    if (selectedList.length < 1) {
      alert("You need to select at least 1 skill");
    } else {
      if (user) {
        await setDoc(
          doc(db, "users", user.uid),
          { focus: selectedList },
          { merge: true }
        );
        alert("User focus skills updated successfully");
      }
    }
  };

  const focus = useMemo(() => {
    if (loading || error) return null;

    return userData?.data()?.focus;
  }, [loading, error, userData]);

  return (
    <Flex flexDirection="column" flexFlow="row wrap">
      <Navbar />
      <Sidebar />
      <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>
        <Box maxW="1350px" style={{ width: "calc(100% - 270px)" }}>
          <RenderWithInitiallySelected focus={focus}>
            {({ initiallySelected }) => (
              <CheckboxGroup
                colorScheme="blue"
                defaultValue={initiallySelected}
              >
                <Wrap>
                  {sIds.length > 0 &&
                    sIds.map((sId) => (
                      <Button
                        as={Checkbox}
                        key={sId}
                        value={sId}
                        onChange={handleCheck}
                        isChecked={selectedMap.get(sId)}
                      >
                        {skillsRoaster.get(sId)}
                      </Button>
                    ))}
                </Wrap>
              </CheckboxGroup>
            )}
          </RenderWithInitiallySelected>
          <Button>Confirm</Button>
        </Box>
      </Box>
    </Flex>
  );
};
```

Notice that I did this bit:

```tsx

  const focus = useMemo(() => {
    if (loading || error) return null;

    return userData?.data()?.focus;
  }, [loading, error, userData]);
```

Just to avoid creating another function, or having a very long ternary.

**Reply by EhanDuan** [reply]
@icyJoseph Your code perfectly solves my doubt, thanks! One last thing: Am I making an easy problem over-complicated? Like we can achieve this purpose in a more standard and cleaner way?

**Reply by icyJoseph** [reply]
Well, it depends.

The userData, which depends on the user, both being async work, so in some capacity you have to handle, the pending, resolve and reject status of both. That's a lot of state, and rolling your own solution, will have some complexity.

Some use `swr` or `react-query`, to do conditional fetching, in addition to the fact that these libraries handle, the pending, resolve, reject, caching, deduping, etc., for you.

I don't know though, how much effort would it take to integrate these with firebase and such, there's probably libraries and what not.

I think, proper component breakdown, like the render props I suggested, can help out to have clear responsibility, on which components fetch data, and which just render it. This is a waterfall pattern, which is not very nice, and that conditional fetching + Suspense and others, are trying to solve. 

You are in the good path I'd say. Just try to judge which state and data-structures should you keep, in order to help your future self understand this when you come back to it.

I kinda wanted to take out the selected Map as state, in favour of derived state, because after you submit, you get an updated version, if I understand firebase reactivity correctly, and the checkboxes just need to know if they are checked, but I get confused here, because I am not sure I understand how `handleChecked` works, at least in these snippets.

**Reply by EhanDuan** [reply]
Thanks for your detailed explanations. I will go to check `swr` and `react-query`. 

The checkboxGroup I am using is from [Chakra UI](https://chakra-ui.com/docs/components/checkbox/usage#checkboxgroup). Currently, I only know how to use it in a simple way: **write a handler and put it to `onChange`**, while I do not fully understand the deeper level of how they realize it. Still need to learn to understand source code.

Thanks again!
